Add ECL lexer

dehilsterlexis opened the pull request:
Adding a lexer for ECL. ECL is the "Enterprise Control Language" for the open source HPCC-System (High Performance Computing Cluster) from LexisNexis.

dabittan commented:
Well done @dehilsterlexis ! Can't wait for this to be merged and reflected in GitLab.

GordonSmith commented on lib/rouge/demos/ecl:
```diff
@@ -0,0 +1,59 @@
+    boolean   B1;
+    decimal23 D1;
+    udecimal3 D2;
+END;
```
@dehilsterlexis can you add a nested RECORD as well?

Also add a MODULE,  MODULEMACRO and EMBED samples?

GordonSmith replied:
(I don't expect the contents of and EMBED to be colored correctly)

dehilsterlexis replied:
This was more of a placeholder for submitting this in a timely manner. I am currently working on a full test file which I will submit.

pyrmont replied:
This is actually overly long for the demo file. That file is only intended to be a short example of the language. Anything more substantial should be in the visual sample. Could you cut this down a bit?

dehilsterlexis replied:
@GordonSmith you are right about the EMBED etc. Fixing now.

dehilsterlexis replied:
@pyrmont yea. I got this wrong. I simply copied my test file. I will replace this with our short "standard" example.

GordonSmith commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+        rule %r(\b(?i:(false|true))\b), Name
+
+        rule %r([\[\]{}();,]), Punctuation
+	rule %r(#{id}), Name
```
Minor: indentation issues

GordonSmith replied:
@dehilsterlexis bump

dehilsterlexis replied:
Already corrected in second commit

pyrmont replied:
Still outstanding.

GordonSmith commented on spec/lexers/ecl_spec.rb:
```diff
@@ -0,0 +1,27 @@
+
+    it 'recognizes one-line comments not followed by a newline (#796)' do
+      assert_tokens_equal '// comment', ['Comment.Single', '// comment']
+    end
```
I suspect you will need full coverage of each category / pattern within the tests?

pyrmont replied:
We don't actually mind if the tests are a bit light. Comprehensively testing a syntax highlighter is pretty difficult so if you just want to use the visual sample to eyeball it, that's acceptable from the perspective of the maintainers.

dehilsterlexis replied:
Don't worry. I am doing it more for our internal use. We submit more than one lexer and it would be helpful for our other lexers. I will push an update later.

pyrmont replied:
No worries. If you want an example of relatively thorough tests, [the GHC Core spec](https://github.com/rouge-ruby/rouge/blob/a821e3f2f5b135345ab51c14ae75002a39eb2cf3/spec/lexers/ghc_core_spec.rb) is pretty good.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+      mimetypes 'application/x-ecl'
+
+      title "ECL"
+      desc "Enterprise Control Language"
```
Is there a URL that can be included in the description for reference (see [the Rust lexer](https://github.com/rouge-ruby/rouge/blob/22ac67833d29a1996ba4b3d3f62577a15eabec2d/lib/rouge/lexers/rust.rb), for example).

dehilsterlexis replied:
Added url

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+      )
+
+      state :string_character_escape do
+        rule %r/\\\\(x\\h{2}|[0-2][0-7]{,2}|3[0-6][0-7]?|37[0-7]?|[4-7][0-7]?|.|$)/, Text
```
Is the double backslash correct? So an escaped string character could be `\\x\h\h`? Separately, a better token for this rule would be `String::Escape`.

dehilsterlexis replied:
You are right. I was using another lexer as an example and it's regex patterns are in double quotes.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+
+      state :single_quote do
+        rule %r([xDQUV]?'([^'\\]*(?:\\.[^'\\]*)*)'), Str::Single
+        mixin :string_character_escape
```
If you're only going to use the `:string_character_escape` state once, it'd be better to inline the rule here (again, just to cut down on the number of objects that are created).

dehilsterlexis replied:
Makes sense

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+
+      state :whitespace do
+        rule %r/\n+/m, Text
+        rule %r(//.*?$), Comment::Single
```
Ruby will match to the end of the line here naturally. You can simplify this to `%r(//.*)`.

dehilsterlexis replied:
Got it.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+      state :inline_whitespace do
+        rule %r/[ \t\r]+/, Text
+        rule %r/\\\n/, Text # line continuation
+        rule %r(/(\\\n)?[*].*?[*](\\\n)?/)m, Comment::Multiline
```
So a multiline comment can look like:

```
/
* This is a comment. * /
```

GordonSmith replied:
Nope that is not permitted (@dehilsterlexis)

dehilsterlexis replied:
Will correct. @GordonSmith

GordonSmith replied:
@dehilsterlexis bump

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+
+      state :root do
+	mixin :whitespace
+	mixin :single_quote
```
These lines should be indented.

GordonSmith replied:
@dehilsterlexis bump

dehilsterlexis replied:
Second commit.

pyrmont replied:
I'm not sure what that means but this remains outstanding.

pyrmont replied:
Indentation issue still here.

dehilsterlexis replied:
@pyrmont fixed. Sorry. Using vi...

dabittan replied:
@pyrmont, please resolve if fixed.

dehilsterlexis replied:
@pyrmont it was fixed 5 days ago.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+        rule %r(\b(?i:(u)?decimal)(\d+(_\d+)?)\b), Keyword::Type
+
+        rule %r(\b(?i:(and|not|or|in))\b), Operator::Word
+        rule %r([:=||>|<|<>|/|\\|+|-|=]), Operator 
```
Are the `|` characters supposed to be interspersed here? You don't need to do that with a character class. Please note because of the meaning `-` can have inside a character class, it should be the last character in a class if intended to match literally.

dehilsterlexis replied:
There is one operator in ECL that is ":=".

GordonSmith replied:
@dehilsterlexis Looks like you have and additional "|" at the start which is not needed:
```
  rule %r([:=||>|<|<>|/|\\|+|-|=]),
```
Should be:
```
  rule %r([:=|>|<|<>|/|\\|+|-|=]),
```
?

dehilsterlexis replied:
Good catch.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+        rule %r(\b(?i:(and|not|or|in))\b), Operator::Word
+        rule %r([:=||>|<|<>|/|\\|+|-|=]), Operator 
+
+        rule %r(\d+\.\d+(e[\+\-]?\d+)?), Num::Float
```
Can the exponent be denoted by both `E` and `e`?

dehilsterlexis replied:
You are correct, they can be both. Will correct.

GordonSmith replied:
@pyrmont that is correct.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+        rule %r([:=||>|<|<>|/|\\|+|-|=]), Operator 
+
+        rule %r(\d+\.\d+(e[\+\-]?\d+)?), Num::Float
+        rule %r(x[0-9a-fA-F]+), Num::Hex
```
You can simplify this regex by using the `i` modified and making it instead `%r(x[0-9a-f]+)i`.

dehilsterlexis replied:
Yep. Better!

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,140 @@
+        rule %r(0[bB][01]+), Num::Bin
+        rule %r([01]+[bB]), Num::Bin
+        rule %r(\d+), Num::Integer
+        rule %r(\b(?i:(false|true))\b), Name
```
The token here should be `Keyword::Constant`.

pyrmont commented on spec/lexers/ecl_spec.rb:
```diff
@@ -0,0 +1,27 @@
+
+    it 'guesses by filename' do
+      assert_guess :filename => 'foo.ecl'
+      assert_guess :filename => 'FOO.ECL'
```
You don't need to check this.

dehilsterlexis replied:
@pyrmont I'm not having luck matching everything between words for our "embed" function. The regex doesn't do that from what I can see. Is there a way to **mixin** languages like Python, c++, etc? This doesn't work. It matches to the last ENDEMBED in the file:

`EMBED[\s\S]*ENDEMBED;`

Our syntax is:

```
INTEGER addone(INTEGER p) := EMBED(Python)
# Python code that returns one more than the value passed to it\n
if p < 10:\n
    return p+1\n
else:\n
    return 0\n
ENDEMBED;
```

Suggestions?

jneen replied:
What you probably want is the `delegate` function. And in general, when writing regexes in a lexer, you want every `*` or `+` to be followed by a `?` - this makes it non-greedy and will match the shortest possible string.

jneen replied:
See for example the HTML lexer, which embeds both CSS and Javascript, for an example of `delegate`.

dehilsterlexis replied:
Thanks!

jneen replied:
Are those `\n`s actually in the source, or is that just an indicator that there are newlines there?

dehilsterlexis replied:
Sorry, shouldn't be there. Fixed.

pyrmont replied:
Just a reminder that you don't need to check the different cases. This line can be deleted.

dehilsterlexis commented:
I have made corrections, updates and additions to the files. One thing I did not solve was the EMBED but will do that in a later push. I believe I have addressed all the comments.

dehilsterlexis commented:
@GordonSmith for review.

dehilsterlexis commented:
Changes made ready for review

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,136 @@
+      ws = %r((?:\s|//.*?\n|/[*].*?[*]/)+)
+      id = /[a-zA-Z_][a-zA-Z0-9_]*/
+
+      template = %w(
```
We generally encourage the memoisation of these attributes. You can see an example of doing that in the JavaScript lexer:

https://github.com/rouge-ruby/rouge/blob/39b0ed7117a421fea1cd1403e8a4e17844653bc7/lib/rouge/lexers/javascript.rb#L99-L105

dehilsterlexis replied:
I tried that in the very beginning but could not get it to work. I will try once again.

pyrmont replied:
The name of the method is written differently if it's within a block passed to `rule` (as was the case in the JavaScript example) than if it's within the rule itself. If it's in a block, you write it as `self.class.<method_name>` but if it's within the rule (as you currently use the local variables), you need to write it as `<LexerClass>.<method_name>`. If

dehilsterlexis replied:
Appreciated!

dehilsterlexis replied:
@pyrmont I'm having problem with case. From my various tests, it seems like the "include" function in memoisation's `"xxx.include? m[0]"` is  is case sensitive. ECL is not case sensitive. The reason I didn't use memoisation is that I don't want to put both cases into the list. Do you know of a way to make "include" in this case not case sensitive? Is there something on the `%w(` that can be applied? I'm not finding anything.

pyrmont replied:
In the block do this:

```ruby 
name = m[0].downcase
if self.class.keywords.include? name
  # ...
elsif self.class.others.include? name
  # ...
end
```

dehilsterlexis replied:
Didn't find that anywhere. I will try that.

pyrmont replied:
It looks like you've memoised most of the variables except for `id`, `template` and `typed`. I understand why `id` would remain but is there any reason for `template` and `typed`?

dehilsterlexis replied:
The templates start with "#".

dabittan replied:
@pyrmont, are there still any pending changes necessary here?

pyrmont replied:
The `#` that's in regex pattern in the template rule is part of Ruby's interpolation syntax. If templates are supposed to start with a `#` then you need an additional `#` to start off the rule, like this:

```ruby
rule %r(#\b(#{template.join('|')})\b)i, Keyword::Type
```

dehilsterlexis replied:
Good catch! Somehow I lost that change!

dehilsterlexis replied:
@pyrmont fixed.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,136 @@
+	mixin :single_quote
+
+        rule %r(\b(?i:(begin|end)c\+\+)), Keyword
+        rule %r(\b(?i:(#{keywords.join('|')}))\b), Keyword
```
Rather than have a large number of similar rules that match against various words, we generally encourage a single rule where the match is checked using Ruby's Array/Set inclusion predicates. You can see how this can work in the JavaScript lexer as well:

https://github.com/rouge-ruby/rouge/blob/39b0ed7117a421fea1cd1403e8a4e17844653bc7/lib/rouge/lexers/javascript.rb#L176-L192

dehilsterlexis replied:
Yes. I was aware of that. I was having problem with including words like "beginc++" with special characters. Now that I have separated them, I can do that.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,136 @@
+	mixin :whitespace
+	mixin :single_quote
+
+        rule %r(\b(?i:(begin|end)c\+\+)), Keyword
```
So this is intended to match `beginc++` and `endc++`?

dehilsterlexis replied:
Yes. In a future push, I will be using "delegate" for embedded languages into ECL.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,136 @@
+        rule %r(\b(?i:(u)?decimal)(\d+(_\d+)?)\b), Keyword::Type
+
+        rule %r(\b(?i:(and|not|or|in))\b), Operator::Word
+        rule %r([:=|>|<|<>|/|\\|+|-|=]), Operator 
```
I'm not sure I made the point very well in the earlier comment.

You've put the operators inside square brackets meaning that everything in between them is treated as members of a character class. In this case `|` is not treated as matching one or the other, it's matching the literal `|` character. To do what you want here you just need to do this: `[:=></\\+-]` (the `-` goes at the end because it otherwise has a special meaning inside a character class). 

You may want to add a `+` after the closing square bracket so that you can match multiple operators in a single pass (although that's not necessary).

GordonSmith replied:
@pyrmont we would need `:=` and `<>` to be evaluated as a single operator...

pyrmont replied:
@GordonSmith Checking syntax correctness is a non-goal for Rouge and so, from our perspective, it's not a problem if you lex characters as tokens that are invalid syntactically. You could achieve what I think you want by having a regex that's `%r(:=|>|<|...)` but that's inconsistent with how the other lexers are written and is using Rouge more as a parser than a lexer.

GordonSmith replied:
@pyrmont understood - I think "need" is the wrong term, its more consistency, if the following should not be colored as a comment:
```
/ *  some comments * /
```
Then the same should be true for `:=`, IOW the following should not color it is an operator:
```
xxx :  = yyy;
```
or am I misunderstanding your original remark?

BTW Thanks for your detailed reviews in all of this as we don't have any ruby experience (and I only have minor regex experience).

Finally - do prefer squashed commits during review?

pyrmont replied:
Sorry—I haven't finished thinking about this. I was doing some maintenance of other lexers yesterday and noticed they do define set operators so am not so certain of my opinion here.

In terms of squashing commits, definitely prefer that you don't do this! What you've been doing so far has been great; for review purposes it's very helpful to be able to compare individual commits! I'll do a squash when I merge it into master.

dabittan replied:
@pyrmont, Please resolve if this discussion is complete.

pyrmont replied:
@dabittan I've changed my mind and think that it's fine to specify precisely the acceptable operators but the way regex pattern that's currently being used isn't the way to do it. Characters in a class (i.e. that are in inside '[' and ']' are matched individually. To achieve what you want, you need to write something like this:

```ruby
rule %r(:=|>|<|<>|/|\\|+|-|=), Operator
```

dabittan replied:
@dehilsterlexis, please let us know when you complete these changes

dehilsterlexis replied:
@dabittan @pyrmont completed the changes.

pyrmont replied:
Sorry, I forgot to escape the `+`. It should be:

```ruby
rule %r(:=|>|<|<>|/|\\|\+|-|=), Operator
```

dehilsterlexis replied:
@pyrmont `rule %r(:=|>|<|<>|/|\\|+|-|=), Operator` does not work. It works with the brackets so I changed it back.

pyrmont replied:
Sorry, I sent it a bit later but did you see my comment? The `+` needed to be escaped.

dehilsterlexis commented:
@pyrmont for review

pyrmont commented on spec/visual/samples/ecl:
```diff
@@ -0,0 +1,146 @@
+    time(&timeinsecs);
+    localtime_r(&timeinsecs,&localt);
+    char temp[15];
+    strftime(temp , 15, "%Y%m%d%H%M%S", &localt); // Formats the localtime to YYYYMMDDhhmmss
```
When I look at this in the visual sample app, the `"` characters are being tokenised as errors.

GordonSmith replied:
`BEGINC++` / `ENDC++` is equivalent to `EMBED(c++)` / `ENDEMBED` so everything between should be ignored now

pyrmont commented on spec/visual/samples/ecl:
```diff
@@ -0,0 +1,146 @@
+/****************** BLOCKS (not working currently) ******************/
+
+INTEGER addone(INTEGER p) := EMBED(Python)
+# Python code that returns one more than the value passed to it
```
Are comments allowed to begin with a `#`? This is tokenising as an error in the visual sample app.

GordonSmith replied:
Everything between `EMBED` and `ENDEMBED` should be ignored now.
This construct is for embedding foreign languages we see colouring those languages as future work.

dehilsterlexis commented:
@pyrmont for review.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,152 @@
+        rule %r([\[\]{}();,\&,\.,\%]), Punctuation
+
+        rule %r(\b(?i:(begin|end)c\+\+)), Keyword
+        rule %r(\b(?i:(std)\.(#{class1.join('|')})\.(#{class2.join('|')}))\b), Name::Class
```
Just out of curiosity, it seems based on the lexer that you can't define your own classes; there are simply the ones defined in the language. Does that mean `A.B.C` is invalid syntax if it doesn't match the `std` classes?

GordonSmith replied:
The language doesn't support classes as such, but it does support nested constructs.  So A.B.C could be valid, but should really only be treated as a variable name.
`Std` is our builtin standard library so gets some special treatment.

GordonSmith commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,152 @@
+        mixin :single_quote
+
+        rule %r(\b(?i:(and|not|or|in))\b), Operator::Word
+        rule %r([:=|>|<|<>|/|\\|+|-|=]), Operator 
```
@dehilsterlexis this should be `rule %r(:=|>|<|<>|/|\\|\+|-|=), Operator` as per @pyrmont comment

dehilsterlexis replied:
@GordonSmith @pyrmont done

dehilsterlexis commented:
@pyrmont for final review

dehilsterlexis commented:
@pyrmont @GordonSmith I have pushed a new version addressing all the issues and added embedding.

When I was adding embedding I found what I think to be a possible problem with the lexer. I was looking to colorize the embedding keywords and the stuff in between as a string color and encountered a case where the lexer added a word to the text. Here is how you can recreate it. The part that is commented out is what I would want to really use instead of the line above:

```
        rule %r(\b(?i:(beginc\+\+.*?endc\+\+)))m, Str::Single
        rule %r(\b(?i:(embed.*?endembed)))m, Str::Single
        #rule %r(\b(?i:((embed)(.)*?(endembed))))m do
          #groups Keyword, Name::Other, Keyword
        #end
```

The output is here. You can see it added the first token m[0] at the end of the matched phrase and colorized is as if it was the first token. It literally inserts new text. Perhaps I am doing something wrong?

![RubyLexerCuriosity](https://user-images.githubusercontent.com/10133951/74357530-0dc60f00-4d8e-11ea-8bd2-64843b4f4ec0.jpg)

dehilsterlexis commented:
@pyrmont nudge

pyrmont commented:
@dehilsterlexis I've pushed some changes to the PR. Hopefully the commit messages are self-explanatory but the key change is to avoid regex interpolation as much as possible by using more generic regular expressions. Hopefully it all makes sense.

> When I was adding embedding I found what I think to be a possible problem with the lexer. I was looking to colorize the embedding keywords and the stuff in between as a string color and encountered a case where the lexer added a word to the text. 

I think this was because you had a nested pattern. I've fixed this in the most recent commit. If you want to use a different lexer for the embedded text, [`RegexLexer#delegate`](https://rouge-ruby.github.io/docs/Rouge/RegexLexer.html#delegate-instance_method) is the method you want. You can see some examples of how it's used in other lexers [here](https://github.com/rouge-ruby/rouge/search?q=delegate&unscoped_q=delegate).

GordonSmith commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -118,7 +118,10 @@ def self.typed
         rule %r(\b(?i:(beginc\+\+.*?endc\+\+)))m, Str::Single
-        rule %r(\b(?i:(embed.*?endembed)))m, Str::Single
+        rule %r(\b(embed)([(])(.*?)([)])(.*?)(endembed))im do |m|
+          groups Keyword, Punctuation, Name::Other, Punctuation, Name::Other, Keyword
```
For the first version I don't we should attempt to colorize anything within the `embed` / `endembed` sections (the content of these sections are other languages: Python, c++, Java, JavaScript, R...) and are only used in "rare" situations.

pyrmont replied:
@GordonSmith I don't mind either way. The rule was in the version of the lexer I edited but please feel free to remove and push.

pyrmont commented on lib/rouge/lexers/ecl.rb:
```diff
@@ -0,0 +1,172 @@
+
+      class2 = %w(
+        debug email job log thorlib util workunit
+      )
```
@dehilsterlexis Why did you revert these to be local variables?

dehilsterlexis replied:
For the root rule which uses them:
`rule %r(\b(?i:(std)\.(#{class1.join('|')})\.(#{class2.join('|')}))\b), Name::Class`

pyrmont replied:
Right but for the reasons explained previously, we try to avoid interpolations in regular expressions. The code I wrote avoided the need to do this. I don't understand what the problem with it was.

dehilsterlexis replied:
@pyrmont I do. But my previous solution was to remove it. Is there another way to do the pattern matching not using variables?

pyrmont replied:
I'm not sure I follow your comment. The code you've changed it to uses both local variables and regex interpolation. I'm not sure what was broken about my approach.

dehilsterlexis replied:
@pyrmont I will try it with regex interp.

pyrmont replied:
We may be talking cross-purposes. We want to _avoid_ regex interpolation. The code I wrote below does that:

```ruby
rule %r(\b(\w+)\.(\w+)\.(\w+)) do |m|
  if m[1] == "std" &&
    self.class.class_first.include?(m[2]) &&
    self.class.class_second.include?(m[3])
    token Name::Class
  else
    token Name::Variable
  end
end
```

pyrmont replied:
(That code is still in the lexer, starting at line 20. Incidentally, it will break now because the `self.class.class_first` and self.class.class_second` methods have been removed.)

dehilsterlexis replied:
@pyrmont I did not see you had put in the code already. My bad. I will put it back.

dehilsterlexis replied:
@pyrmont restored.

pyrmont commented:
@ddehilster I'm ready to merge this. Is there anything left on your side?

dehilsterlexis commented:
@pyrmont we are good to go from our end

pyrmont commented:
@dehilsterlexis Merge successful! We're on a monthly release cadence and the next version is scheduled to ship on Tuesday 10 March. This will be part of that release. Thanks for the contribution!

dabittan commented:
@pyrmont , is the ECL lexer live yet? we still don't see it reflected in our GitLab ecl code.

pyrmont commented:
@dabittan The ECL lexer is part of Rouge [v3.17.0](http://rouge.jneen.net/v3.17.0/ecl/). I'm not associated with GitLab and don't know what version of Rouge they are running.
